cmd/juju/commands/repl.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package commands
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/chzyer/readline"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/loggo"
+
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/jujuclient"
+)
+
+type replCommand struct {
+	cmd.CommandBase
+
+	store    jujuclient.ClientStore
+	showHelp bool
+
+	execJujuCommand func(cmd.Command, *cmd.Context, []string) int
+}
+
+func newReplCommand(showHelp bool) cmd.Command {
+	return &replCommand{
+		showHelp:        showHelp,
+		store:           jujuclient.NewFileClientStore(),
+		execJujuCommand: cmd.Main,
+	}
+}
+
+const replDoc = `
+When run without arguments, enter an interactive shell which can be
+used to run run any Juju command directly.
+Type q|Q or ^D or ^C to quit.
+Otherwise, the supported command usage is described below.
+`
+
+func (c *replCommand) Info() *cmd.Info {
+	return jujucmd.Info(&cmd.Info{
+		Name:    "juju",
+		Purpose: "Enter an interactive shell for running Juju commands",
+	})
+}
+
+func filterInput(r rune) (rune, bool) {
+	switch r {
+	// block CtrlZ feature
+	case readline.CharCtrlZ:
+		return r, false
+	}
+	return r, true
+}
+
+func (c *replCommand) Run(ctx *cmd.Context) error {
+	if err := c.Init(nil); err != nil {
+		return errors.Trace(err)
+	}
+	if c.showHelp {
+		jujuCmd := NewJujuCommand(ctx, "")
+		f := gnuflag.NewFlagSet(c.Info().Name, gnuflag.ContinueOnError)
+		f.SetOutput(ioutil.Discard)
+		jujuCmd.SetFlags(f)
+		if err := jujuCmd.Init([]string{"help"}); err != nil {
+			return errors.Trace(err)
+		}
+		fmt.Fprintln(ctx.Stdout, replDoc)
+		return jujuCmd.Run(ctx)
+	}
+
+	history, err := ioutil.TempFile("", "juju-repl")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer history.Close()
+
+	l, err := readline.NewEx(&readline.Config{
+		Stdin:               readline.NewCancelableStdin(ctx.Stdin),
+		Stdout:              ctx.Stdout,
+		Stderr:              ctx.Stderr,
+		HistoryFile:         history.Name(),
+		InterruptPrompt:     "^C",
+		HistorySearchFold:   true,
+		FuncFilterInputRune: filterInput,
+		// TODO(wallyworld) - add auto complete support
+		//AutoComplete:    jujuCompleter,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer l.Close()
+
+	// Record the default loggo writer so we can
+	// reset it before each command invocation.
+	defaultWriter, err := loggo.RemoveWriter(loggo.DefaultWriterName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		// loggo maintains global state so reset before each command.
+		loggo.ResetLogging()
+		_ = loggo.RegisterWriter(loggo.DefaultWriterName, defaultWriter)
+
+		jujuCmd := NewJujuCommand(ctx, "")
+		if c.showHelp {
```

Comment:
I think the default prompt initial output should at the bare minimum tell the user they can type "help" for commands, and tell them how to exit.

Reply:
Good idea